// infuse/app.js

var Proteus = require('proteus'),
    infuse  = require('infuse'),
    FS      = require('fs')
;

module.exports = new (Proteus.Class.derive(Object.defineProperties({
    
    run: function () {
        var opts = this.options,
            infuseOpts,
            infuser,
            outstream,
            result;
        
        // console.log(opts);
        
        infuseOpts = {
            file:        opts.infile,
            comments:    opts.comments,
            embed:       opts.embed,
            infusions:   opts.infusions,
            modules:     opts.modules,
            nodelib:     opts.nodelib,
            definitions: opts.definitions,
            requires:    opts.requires
        };
        
        if (opts.preProcess) {
            
        }
        
        infuser = infuse(infuseOpts);
        
        outstream = opts.outfile ?
            FS.createWriteStream(opts.outfile) :
            process.stdout;
            
        outstream.setEncoding('utf8');
        
        if (opts.dumpAst) {
            result = JSON.stringify(infuser.ast, null, 4);
        }
        else if (opts.dumpModules) {
            result = opts.requires ? infuser.get('requires').requires : [];
            result = result.join('\n') + '\n';
        }
        else {
            result = infuser.ast.source;
        }
        
        outstream.write(result);
    }
    
}, {
    
    options: {
        get: function () {
            var opts = require('./options');
            Object.defineProperty(this, 'options', { value: opts, enumerable: true });
            return opts;
        },
        enumerable: true
    }
    
})))();
