// infuse/cli-app.js

var Proteus = require('proteus'),
    infuse  = require('infuse'),
    FS      = require('fs')
;

module.exports = new (Proteus.Class.derive(Object.defineProperties({
    
    run: function () {
        var opts = this.options,
            ast,
            outstream,
            result;
        
        // console.log(opts);
        
        result = infuse({
            file:        opts.infile,
            comments:    opts.comments,
            embed:       opts.embed,
            infusions:   opts.infusions,
            modules:     opts.modules,
            nodelib:     opts.nodelib,
            definitions: opts.definitions,
            requires:    opts.requires,
            processor:   opts.preProcess,
            minifier:    opts.minify,
            dump:        opts.dumpAst ? 'ast' : opts.dumpModules ? 'modules' : null
        });

        outstream = opts.outfile ?
            FS.createWriteStream(opts.outfile) :
            process.stdout;
            
        outstream.setEncoding('utf8');
        
        if (opts.dumpAst) {
            result = JSON.stringify(result);
        }
        else if (opts.dumpModules) {
            result = result.join('\n') + '\n';
        }
        
        outstream.write(result || '');
    }
    
}, {
    
    options: {
        get: function () {
            var opts = require('./options');
            Object.defineProperty(this, 'options', { value: opts, enumerable: true });
            return opts;
        },
        enumerable: true
    }
    
})))();
