// infuse/index.js

var Proteus     = require("proteus"),
    Infuser     = require("./infuser"),
    Ast         = require("./ast");

function preProcess (file, processor) {
    
}
/**
 * @param {Object} opts
 *      @property {String}  opts.file file to infuse
 *      @property {Boolean} opts.comments keep comments?
 *      @property {Boolean} opts.embed embed modules as strings?
 *      @property {Boolean} opts.requires process requires?
 *      @property {Array}   opts.infusions
 *      @property {Array}   opts.modules
 *      @property {Array}   opts.nodelib
 *      @property {Object}  opts.definitions
 *      @property {String}  opts.processor processor to run on file before parsing
 *      @property {String}  opts.minifier minifier to use, if any
 * @returns {infuse.Ast}
 */
function infuse (opts) {
    var file = opts.file,
        ast = new Ast(),
        infuser = new Infuser(),
        parserOpts,
        ReqInfusion,
        reqModListKey,
        DefInfusion,
        instance,
        src;
    
    if (opts.comments) {
        // yes, you have to set all these options to get comments
        (parserOpts = ast.options.parse).comment =
            parserOpts.tokens = parserOpts.range = true;
    }
    
    ast.file = file;
    
    if (opts.processor) {
        // we set the source so we operate on the pre-processed source,
        // and not read from the file itself
        ast.source = preProcess(file, opts.processor);
    }
    else if ((src = preProcess(file))) {
        ast.source = src;
    }
    
    if (opts.requires) {
        ReqInfusion = require('./infusions/requires');
        instance = new ReqInfusion({
            nodelib: opts.nodelib,
            parse: parserOpts
        });
        infuser.use(instance);
        reqModListKey = instance.REQUIRED_MODULES_LIST_KEY;
        if (opts.modules) {
            // TODO: add the modules
        }
    }
    
    if (opts.definitions) {
        DefInfusion = require('./infusions/definitions');
        infuser.use(new DefInfusion(opts.definitions));
    }
    
    if (opts.infusions) {
        opts.infusions.forEach(function (name) {
            var Infusion = require(name);
            infuser.use(
                name,
                typeof Infusion === 'function' ? new Infusion() : Infusion
            );
        });
    }
    // minify before embedding...
    if (opts.minifier) {
        // MinInfusion = require('./infusions/minify');
        // infuser.use('minifier', new MinInfusion(opts.minifier));
    }
    
    if (opts.embed) {
        // EmbInfusion = require('./infusions/embed');
        // infuser.use('embedder', new EmbInfusion());
    }
    
    infuser.run(ast);
    
    if (opts.dump) {
        switch (opts.dump) {
            case 'ast':
                return ast;
            case 'modules':
                return (infuser.get(reqModListKey) || []).map(function (mod) {
                    return mod.filepath;
                });
        }
    }
    else {
        return ast.source;
    }
}
//---------------------------------------------------------------------------
// Exports
//---------------------------------------------------------------------------
module.exports = Proteus.merge(infuse, {
    preProcess: preProcess,
    Infuser:    Infuser,
    Infusion:   require('./infusion'),
    Ast:        Ast,
    util:       require('./util')
});
